### PR TITLE
[Master] Fix closure variables related compiler crash inside do clause in query expression

### LINKED ISSUE #36690: Code that broke Ballerina with java.lang.NullPointerException when reading a result-stream of a mysql query
**Description:**
```
[2022-06-20 10:18:58,941] SEVERE {b7a.log.crash} - null 
java.lang.NullPointerException
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.visit(BIROptimizer.java:164)
        at org.wso2.ballerinalang.compiler.bir.model.BIRNode$BIRBasicBlock.accept(BIRNode.java:445)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.lambda$visit$4(BIROptimizer.java:126)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.visit(BIROptimizer.java:126)
        at org.wso2.ballerinalang.compiler.bir.model.BIRNode$BIRFunction.accept(BIRNode.java:403)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.lambda$visit$1(BIROptimizer.java:103)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.visit(BIROptimizer.java:103)
        at org.wso2.ballerinalang.compiler.bir.model.BIRNode$BIRPackage.accept(BIRNode.java:83)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer.optimizePackage(BIROptimizer.java:82)
        at org.wso2.ballerinalang.compiler.bir.BIRGen.genBIR(BIRGen.java:271)
        at io.ballerina.projects.internal.CompilerPhaseRunner.birGen(CompilerPhaseRunner.java:220)
        at io.ballerina.projects.internal.CompilerPhaseRunner.performBirGenPhases(CompilerPhaseRunner.java:157)
        at io.ballerina.projects.ModuleContext.generateCodeInternal(ModuleContext.java:443)
        at io.ballerina.projects.ModuleCompilationState$4.generatePlatformSpecificCode(ModuleCompilationState.java:132)
        at io.ballerina.projects.ModuleContext.generatePlatformSpecificCode(ModuleContext.java:380)
        at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:166)
        at io.ballerina.projects.JBallerinaBackend.<init>(JBallerinaBackend.java:135)
        at io.ballerina.projects.JBallerinaBackend.lambda$from$0(JBallerinaBackend.java:113)
        at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1134)
        at io.ballerina.projects.PackageCompilation.getCompilerBackend(PackageCompilation.java:169)
        at io.ballerina.projects.JBallerinaBackend.from(JBallerinaBackend.java:112)
        at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:93)
        at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
        at io.ballerina.cli.cmd.RunCommand.execute(RunCommand.java:205)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at io.ballerina.cli.launcher.Main.main(Main.java:51)
 ```

**Steps to reproduce:**
In a GET service endpoint, change [1] to [2].
[1]
```
  stream<Quotation, error?> resultQuotationStream = dbClient->query(`SELECT QUOTATIONID, SUPPLIERID, BRANDNAME, AVAILABLEQUANTITY, EXPIRYDATE, UNITPRICE, REGULATORYINFO FROM QUOTATION;`);
  check from Quotation quotation in resultQuotationStream
  do {
        info.supplierQuotes.push(quotation);
  };
  foreach Quotation quotation in info.supplierQuotes {
        Supplier supplier = check dbClient->queryRow(`SELECT SUPPLIERID, NAME, SHORTNAME, EMAIL, PHONENUMBER FROM SUPPLIER WHERE SUPPLIERID=${quotation.supplierID};`);
        quotation.supplier = supplier;
  }
```

[2] 
```
  stream<Quotation, error?> resultQuotationStream = dbClient->query(`SELECT QUOTATIONID, SUPPLIERID, BRANDNAME, AVAILABLEQUANTITY, EXPIRYDATE, UNITPRICE, REGULATORYINFO FROM QUOTATION;`);
  check from Quotation quotation in resultQuotationStream
  do {
        Supplier supplier = check dbClient->queryRow(`SELECT SUPPLIERID, NAME, SHORTNAME, EMAIL, PHONENUMBER FROM SUPPLIER WHERE SUPPLIERID=${quotation.supplierID};`);
        quotation.supplier = supplier;
        info.supplierQuotes.push(quotation);
  };
```
**Affected Versions:** Ballerina SDK 2201.0.4

**OS, DB, other environment details and versions:** Choreo, MYSQL connector

### PR DESCRIPTION
**Pull request by SasinduDilshara**
Fixes #36690 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by LakshanWeerasinghe** [comment]
In the original issue there is a `check` expression inside the `do` action. Can you add a test with a `check` expression inside the `do` action.

`string _ = check str1.ensureType();`

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 1]
```diff
@@ -5949,8 +5949,12 @@ public void visit(BLangSimpleVarRef varRefExpr) {
         // Restore the original symbol
         if ((varRefExpr.symbol.tag & SymTag.VARIABLE) == SymTag.VARIABLE) {
             BVarSymbol varSymbol = (BVarSymbol) varRefExpr.symbol;
+            boolean varRefExprClosure = varSymbol.closure;
```
Comment (ushirask) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:5952]: Can move this inside the next if block.

File: tests/jballerina-unit-test/src/test/resources/test-src/query/query_with_closures.bal [hunk 2]
```diff
@@ -172,6 +172,113 @@ function testClosureInQueryActionInDo2() {
...
+    }
+}
+
+string str1 = "string-1";
+
+function testClosureInQueryActionInDo3() returns error? {
+  string str2 = "string-2";
+  EvenNumberGenerator evenGen = new ();
+  stream<int, error?> evenNumberStream = new (evenGen);
+  int|string str3 = "string-3";
```
Comment (KavinduZoysa) [inline tests/jballerina-unit-test/src/test/resources/test-src/query/query_with_closures.bal:192]: Please check the indentation.

**Comment by SasinduDilshara** [comment]
> In the original issue there is a `check` expression inside the `do` action. Can you add a test with a `check` expression inside the `do` action.
> 
> `string _ = check str1.ensureType();`

updated

File: tests/jballerina-unit-test/src/test/resources/test-src/query/query_with_closures.bal [hunk 3]
```diff
@@ -172,6 +172,116 @@ function testClosureInQueryActionInDo2() {
...
+        }
+        return {value: self.i};
+    }
+}
+
+string str1 = "string-1";
+
+function testClosureInQueryActionInDo3() returns error? {
+    string str2 = "string-2";
+    EvenNumberGenerator evenGen = new ();
```
Comment (LakshanWeerasinghe) [inline tests/jballerina-unit-test/src/test/resources/test-src/query/query_with_closures.bal:190]:
```suggestion
    EvenNumberGenerator evenGen = new;
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 4]
```diff
@@ -5950,7 +5950,11 @@ public void visit(BLangSimpleVarRef varRefExpr) {
         if ((varRefExpr.symbol.tag & SymTag.VARIABLE) == SymTag.VARIABLE) {
             BVarSymbol varSymbol = (BVarSymbol) varRefExpr.symbol;
             if (varSymbol.originalSymbol != null) {
+                boolean varRefExprClosure = varSymbol.closure;
                 varRefExpr.symbol = varSymbol.originalSymbol;
+                if (varRefExprClosure) {
```
Comment (LakshanWeerasinghe) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:5955]:
```suggestion
                varRefExpr.symbol = varSymbol.originalSymbol;
                if (varSymbol.closure) {
```
Can we do this?

File: tests/jballerina-unit-test/src/test/resources/test-src/query/query_with_closures.bal [hunk 5]
```diff
@@ -172,6 +172,116 @@ function testClosureInQueryActionInDo2() {
...
+            string objectName3 = check object3.getName();
+
+            assertEquality("John", objectName1);
+            assertEquality("Jane", objectName2);
+            assertEquality("Anne", objectName3);
+        };
+    };
+  }
+}
+
```
Comment (pcnfernando) [inline tests/jballerina-unit-test/src/test/resources/test-src/query/query_with_closures.bal:284]: Can you add some tests for query expressions, nested query actions and expressions

Comment (SasinduDilshara) [reply tests/jballerina-unit-test/src/test/resources/test-src/query/query_with_closures.bal:284]: will add.

Comment (SasinduDilshara) [reply tests/jballerina-unit-test/src/test/resources/test-src/query/query_with_closures.bal:284]: added

**Review by LakshanWeerasinghe — APPROVED** [review]

**Review by pcnfernando — APPROVED** [review]

**Pull request merged**